ledger/src/blockstore_db.rs

```diff
@@ -1210,6 +1217,28 @@ impl ColumnName for columns::Index {
 }
 impl TypedColumn for columns::Index {
     type Type = blockstore_meta::Index;
+
+    fn deserialize(data: &[u8]) -> Result<Self::Type> {
+        let config = bincode::DefaultOptions::new()
+            // `bincode::serialize` uses fixint encoding by default, so we need to use the same here
+            .with_fixint_encoding()
+            .reject_trailing_bytes();
+
+        // Migration strategy for new column format:
+        // 1. Release 1: Add ability to read new format as fallback, keep writing old format
+        // 2. Release 2: Switch to writing new format, keep reading old format as fallback
+        // 3. Release 3: Remove old format support once stable
+        // This allows safe downgrade to Release 1 since it can read both formats
+        // https://github.com/anza-xyz/agave/issues/3570
+        let index: bincode::Result<blockstore_meta::Index> = config.deserialize(data);
```

Comment:
do we have tests for [de]serialization failing? I know that if you know how it
works it's probably obvious why it fails, but I have pathologically bad memory
and so I just spent 15 minutes rediscovering all the cases for the 3rd time 😅

I think that we should test that:

- deserializing IndexV2 from an empty Index fails because of UnexpectedEof
  (missing data for num_shreds)
- deserializing IndexV2 from a non-empty Index fails because of
  trailing data (Index encodes u64s, IndexV2 encodes u8s)
- deserializing Index from any IndexV2 fails with UnexpectedEof (tries to
  read 4096 u64s when there are only 4096 u8s)

Reply:
Oh I don't know how I missed those sorry! 🤦‍♂️🤦‍♂️🤦‍♂️

The tests are 100% fine, but yes please add some comments that explain _why_ deserialize fails because from is_err() it isn't clear